// @inesat/cli/dist/helper/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.logTag = exports.logTitle = void 0;
const chalk_1 = __importDefault(require("chalk"));
const store_1 = __importDefault(require("./store"));
// const prefix = `[InesatCli]`
const logger = {
    info: (...args) => ['debug', 'info'].includes(store_1.default.config.debug.level) && console.log(...args),
    debug: (...args) => ['debug'].includes(store_1.default.config.debug.level) && console.log(...args),
    warn: (...args) => ['debug', 'info', 'warn'].includes(store_1.default.config.debug.level) && console.warn(...args),
    error: (...args) => ['debug', 'info', 'warn', 'error'].includes(store_1.default.config.debug.level) && console.error(...args),
};
/**
 * begin logger
 * @param title
 * @returns
 */
const logTitle = (title) => console.log(`${chalk_1.default.cyan(`InesatCli v${store_1.default.inesatCliPkg.version}`)} ${chalk_1.default.green(title)} \n`);
exports.logTitle = logTitle;
const logTagStyle = (msg, c1, c2, w = '#ecf0f1') => console.log(`${chalk_1.default.bgHex(w).hex(c1)(` InesatCli v${store_1.default.inesatCliPkg.version} `)}${chalk_1.default.hex(w).bgHex(c2)(` ${msg} `)}\n`);
const logTag = (msg, tag = 'blue') => {
    switch (tag) {
        case 'green':
            logTagStyle(msg, '#27ae60', '#2ecc71');
            break;
        case 'blue':
            logTagStyle(msg, '#2980b9', '#3498db');
            break;
        case 'red':
            logTagStyle(msg, '#c0392b', '#e74c3c');
            break;
        case 'yellow':
            logTagStyle(msg, '#f39c12', '#f1c40f');
            break;
        case 'purple':
            logTagStyle(msg, '#8e44ad', '#9b59b6');
            break;
        case 'black':
            logTagStyle(msg, '#2c3e50', '#34495e');
            break;
    }
};
exports.logTag = logTag;
exports.default = logger;
