// @inesat/cli/dist/webpack/file.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const wpChain_1 = __importDefault(require("../helper/wpChain"));
class WPFile {
    constructor() { }
    async setup() {
        const config = {
            module: {
                rule: {
                    // 解决svgr ReactComponent 无法获取的问题
                    svg: {
                        test: /\.svg$/,
                        // exclude: /(node_modules|bower_components)/,
                        use: {
                            // swc: {
                            //   loader: store.inesatCliResolve(path.resolve(store.inesatCliSource, 'webpack/loader/swc')),
                            //   options: store.config.build
                            // },
                            // ...loader().config,
                            svgr: {
                                loader: require.resolve('@svgr/webpack'),
                                // options: {
                                //   babel: false,
                                // }
                            },
                            url: {
                                loader: require.resolve('url-loader'),
                                options: {
                                    esModule: false
                                }
                            }
                        }
                    },
                    image: {
                        test: /\.(png|jpe?g|gif|webp|ico)$/i,
                        type: 'asset/resource'
                    },
                    fonts: {
                        test: /\.(|otf|ttf|eot|woff|woff2)$/i,
                        type: 'asset/resource'
                    },
                    inline: {
                        resourceQuery: /inline/,
                        type: 'asset/inline'
                    },
                    // 解决 svga 解析失败问题
                    svga: {
                        test: /\.(svga)$/i,
                        type: 'asset/resource'
                    }
                }
            }
        };
        wpChain_1.default.merge(config);
    }
}
exports.default = WPFile;
