// @inesat/cli/dist/webpack/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const store_1 = __importDefault(require("../helper/store"));
const common_1 = __importDefault(require("./common"));
const development_1 = __importDefault(require("./development"));
const file_1 = __importDefault(require("./file"));
const css_1 = __importDefault(require("./css"));
const module_1 = __importDefault(require("./module"));
const plugin_1 = __importDefault(require("./plugin"));
const production_1 = __importDefault(require("./production"));
const entries_1 = __importDefault(require("./entries"));
class WPConfig {
    constructor() {
        this.entries = new entries_1.default();
        this.css = new css_1.default();
        this.common = new common_1.default();
        this.module = new module_1.default();
        this.plugin = new plugin_1.default();
        this.file = new file_1.default();
        this.development = new development_1.default();
        this.production = new production_1.default();
    }
    async setup() {
        await Promise.all([
            this.entries.setup(),
            this.common.setup(),
            this.module.setup(),
            this.plugin.setup(),
            this.file.setup(),
            this.css.setup()
        ]);
        if (store_1.default.config.mode === 'development')
            await this.development.setup();
        else if (store_1.default.config.mode === 'production')
            await this.production.setup();
    }
}
exports.default = new WPConfig();
